Improve module info extraction to gather info safely in parallel

ghale opened the pull request:
This PR adds two things:

- Some rudimentary integration testing for Gradle.  This adds some automation to download an instance of Artifactory open source, start it up before running the integration tests and shut it down at the end of the build.  The integration tests pull their test projects from `src/integTest/resources` and build/publish them to the artifactory instance, validating that the the artifacts and build info look correct.  The test projects are based on the one found [here](https://github.com/jfrog/project-examples/tree/master/gradle-examples/gradle-example-publish).  It is intentionally not hooked into the `check` lifecycle because I'm not 100% sure of the reliability of the start/stop automation and it makes a number of assumptions that might not be valid in all cases (e.g. that it can run on port 8081, etc).  With a little work it could potentially be usable in an overall build pipeline, though.
- Refactor of the module info extraction to make it safer.  With #287 this was parallelized in a fairly unsafe manner using some Gradle internal APIs.  This breaks up the module info extraction so that instead of a single task in the root project reaching across and resolving configurations in other projects, each project has its own extraction task that produces a module info artifact by resolving only the configurations in its project.  Then the root project has a configuration that aggregates all of the module info artifacts and adds them to the build info object.  Since this is done in a more "standard" way from a Gradle perspective, it allows us to remove the internal Gradle API usages.  Since those module info tasks run in different projects, they can safely execute in parallel prior to the deploy task in the root project starting.  This allows us to push some of this work earlier in the build and run it in parallel to other stuff that might be blocking the main deploy task.  YMMV, but in certain cases this could provide a fair performance improvement.

There are still some rough edges and some things that could be improved here but I wanted to get some eyes on it before I go further.  Please take a look and let me know what you think.  I'd be particularly interested in understanding what further test cases would be valuable to add to the integration tests.

ljacomet commented:
@ghale This is great 👍
One thing I would do is use a custom attribute instead of reusing the Gradle `org.gradle.usage` one, just to isolate better the variant identity of the build info.

ghale commented:
Thanks @ljacomet - I've changed it to use a jfrog-specific attribute.

ghale commented:
@yahavi Have you had a chance to look at this at all?  We can split this into two PRs if necessary, too (one for automated tests, one for build info refactor).

ghale commented:
@yahavi Thanks for the feedback.  I've addressed the code comments above.  I tried to implement my test using `IntegrationTestsBase` per your recommendation.  The problem is that this class tries to create repositories in Artifactory via the rest API and according to the docs that function requires Artifactory Pro.  Running the test against Artifactory OSS gives me an error on that step.  

The fundamental problem here is that it's difficult to contribute anything meaningful to this project without some sort of automated regression testing.  In fact, my primary motivation for adding the automated testing in the first place was because I wasn't comfortable doing the refactor in this PR without some regression tests in the first place.

To use the`IntegrationTestsBase` framework, it seems like you 1) have to manually download, start and manage an Artifactory instance and 2) that instance needs to be Artifactory Pro.  To use the Pro version, you need a trial license key which then means you'll be contacted (repeatedly) by the sales team, which is a strong disincentive for contributing.  If you're not working on this project on a regular basis, the chances are you'll have to do this _every time_.  

Ultimately, I think it would be great if there were an automated way to bootstrap the whole download/start/configure step, but even if it were manual, as long as it's painless, I think it's a step in the right direction.  Based on my current understanding, I'm not sure I see a clear path to this with how things are now.

Some questions related to this:
- Is there a way to get an Artifactory Pro license without signing up for a trial?  I think it would be perfectly alright if it expired somewhat aggressively as long as it's easy to get another one for development purposes and would not dump you into the sales pipeline.
- Is there a way to enable the rest services in Artifactory OSS (at least enough to enable the testing automation)?  This would relieve the need for a Pro license.
- Is there another way to implement the test automation in Artifactory OSS without using the rest services?
- Is there something about the current state of things that I'm not understanding?

ghale commented on build-info-extractor-gradle/src/main/groovy/org/jfrog/gradle/plugin/artifactory/extractor/ProjectUtils.java:
```diff
@@ -0,0 +1,48 @@
+        try {
+            return task.getState().getDidWork();
+        } catch (NoSuchMethodError error) {
+            // Compatibility with older versions of Gradle:
```
This was lifted straight out of the existing code.  I think it's left over from some very early Gradle versions.  It certainly is not an issue with the Gradle 4+.  I've just removed this.

ghale commented on build-info-extractor-gradle/src/main/groovy/org/jfrog/gradle/plugin/artifactory/task/ExtractModuleTask.java:
```diff
@@ -0,0 +1,35 @@
+            this.moduleFile = getProject().getObjects().fileProperty();
+        } catch(NoSuchMethodError e) {
+            // Gradle 4.x
+            this.moduleFile = getProject().getLayout().fileProperty();
```
Done.

ghale commented on build-info-extractor/src/main/java/org/jfrog/build/extractor/ModuleExtractorUtils.java:
```diff
@@ -0,0 +1,64 @@
+import java.io.StringReader;
+import java.io.StringWriter;
+
+public class ModuleExtractorUtils {
```
Done.

ghale commented:
@yahavi  I created #301 to carve off the functional changes in this PR so that they can be integrated separately.  I've also created #302 to continue the discussion around integration tests.  I'm happy to take another stab once we have a resolution there.
